share/eds/blockstore.go

```diff
@@ -0,0 +1,210 @@
+
+// Put is a noop on the EDS blockstore, but it does not return an error because it is called by
+// bitswap. For clarification, an implementation of Put does not make sense in this context because
+// it is unclear which CAR file the block should be written to.
+func (bs *Blockstore) Put(context.Context, blocks.Block) error {
+	return nil
+}
+
+// PutMany is a noop on the EDS blockstore, but it does not return an error because it is called by
+// bitswap. For clarification, an implementation of PutMany does not make sense in this context
+// because it is unclear which CAR file the blocks should be written to.
+func (bs *Blockstore) PutMany(context.Context, []blocks.Block) error {
+	return nil
+}
+
+// AllKeysChan is a noop on the EDS blockstore because the keys are not stored in a single CAR file.
+func (bs *Blockstore) AllKeysChan(context.Context) (<-chan cid.Cid, error) {
+	return nil, ErrUnsupportedOperation
+}
+
+// HashOnRead is a noop on the EDS blockstore but an error cannot be returned due to the method
+// signature from the blockstore interface.
+func (bs *Blockstore) HashOnRead(bool) {
+	log.Warnf("HashOnRead is a noop on the EDS blockstore")
+}
+
+// getReadOnlyBlockstore finds the underlying blockstore of the shard that contains the given CID.
+func (bs *Blockstore) getReadOnlyBlockstore(ctx context.Context, cid cid.Cid) (dagstore.ReadBlockstore, error) {
+	keys, err := bs.store.dgstr.ShardsContainingMultihash(ctx, cid.Hash())
+	if err != nil {
+		return nil, fmt.Errorf("failed to find shards containing multihash: %w", err)
+	}
+	if len(keys) != 1 {
+		return nil, ErrInvalidShardCount
+	}
+
+	// try to fetch from cache
+	shardKey := keys[0]
+	blockstr, err := bs.readFromBSCache(shardKey)
+	if err != nil {
+		log.Debugf("blockstore cache miss for key %s: %s", shardKey, err)
+	}
+	if err == nil && blockstr != nil {
+		return blockstr, nil
+	}
+
+	// wasn't found in cache, so acquire it and add to cache
+	ch := make(chan dagstore.ShardResult, 1)
+	err = bs.store.dgstr.AcquireShard(ctx, shardKey, ch, dagstore.AcquireOpts{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize shard acquisition: %w", err)
+	}
+
+	select {
+	case res := <-ch:
+		if res.Error != nil {
+			return nil, fmt.Errorf("failed to acquire shard: %w", res.Error)
+		}
+		return bs.addToBSCache(shardKey, res.Accessor)
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+}
+
+func (bs *Blockstore) readFromBSCache(shardContainingCid shard.Key) (dagstore.ReadBlockstore, error) {
+	lk := &bs.bsStripedLocks[shardKeyToStriped(shardContainingCid)]
+	lk.Lock()
+	defer lk.Unlock()
+
+	// We've already ensured that the given shard has the cid/multihash we are looking for.
+	val, ok := bs.blockstoreCache.Get(shardContainingCid)
+	if !ok {
+		return nil, errors.New("not found in cache")
+	}
+
+	rbs, ok := val.(*accessorWithBlockstore)
+	if !ok {
+		return nil, fmt.Errorf(
+			"casting value from cache to accessorWithBlockstore: %s",
+			reflect.TypeOf(val),
+		)
+	}
```

Comment:
@renaynay debate

Reply:
Should we panic on non-critical execution pass? We can `log.Error` it so node keeps being operational if it happens. I would suggest creating an variable for error:
```
var errNotFound = errors.New("not found in cache")
```
Return it and if returned error is not `errNotFound` log an error.

Reply:
There is nothing to really debate since both are technically fine but, I don't really see a reason to panic here since it isn't critical like @walldiss mentioned.

Reply:
took @walldiss 's suggestion

Reply:
> Should we panic on non-critical execution pass?

Guys, it's critical. If we get the value from the map, which we are the only writers - it's something really wrong.

Reply:
Unfortunately, we are not in the rust world to catch such things during compile time, but we are in the Golang world where we have panics that are used when something impossible happens. This is the case where something impossible happens, and we should panic here, so once it becomes possible due to some new dev coming here and changing, it breaks early locally, in the test network, or the integration test.

Reply:
fair point @Wondertan

Reply:
we now panic